wFOV Immersive MSE content in Fullscreen Immersive view is too close to user

jyavenard opened the pull request:
#### 74f8a14a79115a427f68969a1a3c0e79f0de50d3
<pre>
wFOV Immersive MSE content in Fullscreen Immersive view is too close to user
<a href="https://bugs.webkit.org/show_bug.cgi?id=293573">https://bugs.webkit.org/show_bug.cgi?id=293573</a>
<a href="https://rdar.apple.com/149733671">rdar://149733671</a>

Reviewed by Youenn Fablet.

Copy keys required by REKit to properly process the image in immersive mode.

Canonical link: <a href="https://commits.webkit.org/295444@main">https://commits.webkit.org/295444@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1cb90a6fdbc63438c8d98ce5ee206a2b1bd998de

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/105123 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/24837 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/15270 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/110339 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/55797 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/107164 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/25262 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/33381 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79841 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/108129 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/19685 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94892 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/60148 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/19435 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12968 "Found 2 new test failures: http/tests/security/cookie-module-propagate.html http/tests/security/cookie-module.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/55180 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/89149 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/13012 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/112877 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/32288 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23778 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88919 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/32654 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/91110 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/88549 "Found 101 new API test failures: /TestWebKit:WebKit.CanHandleRequest, /WebKitGTK/TestGeolocationManager:/webkit/WebKitGeolocationManager/watch-position, /TestWebKit:WebKit.UserMediaBasic, /TestWebKit:WebKit.PageLoadEmptyURL, /WebKitGTK/TestOptionMenu:/webkit/WebKitWebView/option-menu-groups, /TestWebKit:WebKit.GetInjectedBundleInitializationUserDataCallback, /TestWebKit:WebKit.GeolocationTransitionToHighAccuracy, /TestWebKit:WebKit.PendingAPIRequestURL, /TestWebKit:WebKit.FirstMeaningfulPaint, /WebKitGTK/TestUIClient:/webkit/WebKitWebView/file-chooser-request ... (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/22574 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/33442 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/11238 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/27694 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/32211 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/37601 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/32001 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/35344 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/33562 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -136,30 +154,80 @@ static bool isNonRecoverableError(OSStatus status)
+            auto height = CVPixelBufferGetHeight(imageBuffer);
+            if (auto status = PAL::CMVideoFormatDescriptionCreate(kCFAllocatorDefault, codecType, width, height, copyExtensions.get(), &rawImageBufferDescription); status != noErr)
+                return makeUnexpected(status);
+            imageBufferDescription = adoptCF(rawImageBufferDescription);
```
Can we move this code to a separate function with a name like `copyDescriptionExtensionValues`?

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -136,30 +154,80 @@ static bool isNonRecoverableError(OSStatus status)
+        static constexpr size_t numberOfKeys = sizeof(keys) / sizeof(keys[0]);
+        size_t keysSet = 0;
+        Vector<RetainPtr<CFPropertyListRef>, numberOfKeys> values(numberOfKeys, [&](size_t index) -> RetainPtr<CFPropertyListRef> {
+            WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
```
Can we use `unsafeMakeSpan` to remove this macro?

youennf commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -136,30 +154,80 @@ static bool isNonRecoverableError(OSStatus status)
+            for (size_t index = 0; index < numberOfKeys; index++) {
+                WTF_ALLOW_UNSAFE_BUFFER_USAGE_BEGIN
+                if (values[index])
+                    CFDictionarySetValue(copyExtensions.get(), keys[index], values[index].get());
```
Why do we even need values here?
Can we just get the value and `CFDictionarySetValue` it iterating over keys as an array or span?

jyavenard replied:
because that way I don't have to allocate or copy a new CMFormatDescription unless necessary

jernoble commented on Source/WebCore/platform/graphics/cocoa/WebCoreDecompressionSession.mm:
```diff
@@ -53,10 +54,25 @@
 
 namespace WebCore {
 
+static bool s_canCopyFormatDescriptionExtension = false;
```
If it were me, I'd have done something like:

```
static bool canCopyFormatDescriptionExtension()
{
    static canCopyFormatDescriptionExtension = PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_HasLeftStereoEyeView()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_HasRightStereoEyeView()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_HeroEye()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_HorizontalDisparityAdjustment()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_StereoCameraBaseline()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_ProjectionKind()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_ViewPackingKind()
        && PAL::canLoad_CoreMedia_kCMFormatDescriptionExtension_CameraCalibrationDataLensCollection();
    return canCopyFormatDescriptionExtension;
}
```

...rather than create a static global variable, but tomato tomahto.

jyavenard committed:
wFOV Immersive MSE content in Fullscreen Immersive view is too close to user
https://bugs.webkit.org/show_bug.cgi?id=293573
rdar://149733671

Reviewed by Youenn Fablet.

Copy keys required by REKit to properly process the image in immersive mode.

Canonical link: https://commits.webkit.org/295444@main
